phpseclib/Crypt/Hash.php

```diff
@@ -213,79 +152,22 @@ function setHash($hash)
                 break;
             case 'sha512':
                 $this->l = 64;
-        }
-
-        switch ($hash) {
-            case 'md2':
-                $mode = CRYPT_HASH_MODE == self::MODE_HASH && in_array('md2', hash_algos()) ?
-                    self::MODE_HASH : self::MODE_INTERNAL;
-                break;
-            case 'sha384':
-            case 'sha512':
-                $mode = CRYPT_HASH_MODE == self::MODE_MHASH ? self::MODE_INTERNAL : CRYPT_HASH_MODE;
                 break;
             default:
-                $mode = CRYPT_HASH_MODE;
-        }
-
-        switch ($mode) {
-            case self::MODE_MHASH:
-                switch ($hash) {
-                    case 'md5':
-                        $this->hash = MHASH_MD5;
-                        break;
-                    case 'sha256':
-                        $this->hash = MHASH_SHA256;
-                        break;
-                    case 'sha1':
-                    default:
-                        $this->hash = MHASH_SHA1;
+                // see if the hash isn't "officially" supported see if it can be "unofficially" supported and calculate the length accordingly
+                if (in_array($hash, hash_algos())) {
+                    $this->l = strlen(hash($hash, '', true));
+                    break;
                 }
-                return;
-            case self::MODE_HASH:
-                switch ($hash) {
-                    case 'md5':
-                        $this->hash = 'md5';
-                        return;
-                    case 'md2':
-                    case 'sha256':
-                    case 'sha384':
-                    case 'sha512':
-                        $this->hash = $hash;
-                        return;
-                    case 'sha1':
-                    default:
-                        $this->hash = 'sha1';
+                // if the hash algorithm doens't exist maybe it's a truncated hash. eg. md5-96 or some such
+                if (preg_match('#(-\d+)$#', $hash, $matches) && in_array($hash = substr($hash, 0, -strlen($matches[1])), hash_algos())) {
+                    $this->l = abs($matches[1]) >> 3;
+                    break;
                 }
-                return;
+                throw new UnsupportedAlgorithmException("$hash is not a supported algorithm");
```

Comment:
Please also add a test case for this. See https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.expectedException